### PR TITLE
packages/kata-kernel: add GPU variant

### PR DESCRIPTION
**Pull request by msanft**
This adds a kernel build with GPU support to the `kata.kata-kernel` package.

### CODE REVIEW

File: packages/by-name/kata/kata-kernel-uvm/package.nix [hunk 1]
```diff
@@ -2,11 +2,13 @@
 # SPDX-License-Identifier: AGPL-3.0-only
 
 {
+  lib,
   fetchurl,
   linuxManualConfig,
   stdenvNoCC,
   fetchzip,
   kata,
+  withGPU ? false,
```
Comment (burgerdev) [inline packages/by-name/kata/kata-kernel-uvm/package.nix:11]: Why do we need a separate kernel for GPUs?

Comment (msanft) [reply packages/by-name/kata/kata-kernel-uvm/package.nix:11]: This is aligned with how Kata handles this upstream. They build minimal kernels for each use-case, which is desirable from both a hardening perspective as well as the storage footprint.

Comment (burgerdev) [reply packages/by-name/kata/kata-kernel-uvm/package.nix:11]: Is the diff actually that big?

File: packages/by-name/kata/kata-kernel-uvm/package.nix [hunk 2]
```diff
@@ -20,28 +22,51 @@ let
...
-        --replace-fail '# CONFIG_NLS_ISO8859_1 is not set' 'CONFIG_NLS_ISO8859_1=y' \
-        --replace-fail '# CONFIG_ATA is not set' 'CONFIG_ATA=y'
+            # Enable dm-init, so that we can use `dm-mod.create`.
+            cat <<EOF >> $config
+            CONFIG_MD=y
+            CONFIG_BLK_DEV_DM=y
+            CONFIG_DM_INIT=y
+            CONFIG_DM_CRYPT=y
+            CONFIG_DM_VERITY=y
+            CONFIG_DM_INTEGRITY=y
```
Comment (burgerdev) [inline packages/by-name/kata/kata-kernel-uvm/package.nix:38]: Don't we want these configs for the regular image, too?

File: packages/by-name/kata/kata-kernel-uvm/package.nix [hunk 3]
```diff
@@ -20,28 +22,51 @@ let
...
-        --replace-fail '# CONFIG_TMPFS_XATTR is not set' 'CONFIG_TMPFS_XATTR=y' \
-        --replace-fail '# CONFIG_EFIVAR_FS is not set' 'CONFIG_EFIVAR_FS=y' \
-        --replace-fail '# CONFIG_RD_ZSTD is not set' 'CONFIG_RD_ZSTD=y' \
-        --replace-fail '# CONFIG_VFAT_FS is not se' 'CONFIG_VFAT_FS=y' \
-        --replace-fail '# CONFIG_NLS_CODEPAGE_437 is not set' 'CONFIG_NLS_CODEPAGE_437=y' \
-        --replace-fail '# CONFIG_NLS_ISO8859_1 is not set' 'CONFIG_NLS_ISO8859_1=y' \
-        --replace-fail '# CONFIG_ATA is not set' 'CONFIG_ATA=y'
+            # Enable dm-init, so that we can use `dm-mod.create`.
+            cat <<EOF >> $config
+            CONFIG_MD=y
```
Comment (burgerdev) [inline packages/by-name/kata/kata-kernel-uvm/package.nix:33]: Is this needed, or is `CONFIG_BLK_DEV_DM` enough?

File: packages/by-name/kata/kata-kernel-uvm/package.nix [hunk 4]
```diff
@@ -20,28 +22,51 @@ let
...
+        else
+          ''
+            config=$(find . -regex '.*/config-[0-9.-]+-confidential')
 
-      echo "CONFIG_ATA_PIIX=y" >> $config
-    '';
+            # 1. Our kernel build is independent of any initrd.
+            # 2. Enable dm-init, so that we can use `dm-mod.create`.
+            substituteInPlace $config \
+              --replace-fail 'CONFIG_INITRAMFS_SOURCE="initramfs.cpio.gz"' 'CONFIG_INITRAMFS_SOURCE=""' \
```
Comment (burgerdev) [inline packages/by-name/kata/kata-kernel-uvm/package.nix:48]: Why is this not needed for the GPU image?

File: packages/by-name/kata/kata-kernel-uvm/package.nix (hunk 1 above)
Comment (msanft) [reply packages/by-name/kata/kata-kernel-uvm/package.nix:11]: I wouldn't know why we should diverge from it, at least. 

And for the size of the diff: I can get you exact numbers, but I wouldn't expect it to be small. Consider driver-disablement (e.g. nouveau), PCI support, etc.

File: packages/by-name/kata/kata-kernel-uvm/package.nix (hunk 4 above)
Comment (msanft) [reply packages/by-name/kata/kata-kernel-uvm/package.nix:48]: The GPU config doesn't have this set in the first place, so we can't replace it.

File: packages/by-name/kata/kata-kernel-uvm/package.nix (hunk 1 above)
Comment (msanft) [reply packages/by-name/kata/kata-kernel-uvm/package.nix:11]: Looking at the diff, we _could_ use the GPU config for both. The real diff is about 20 lines. (not counting moved options, etc.)

However, I'd still prefer to keep close to upstream, as this means that we don't have to keep up with the changes to both files on each Kata release. Do you have a strong opinion @burgerdev @katexochen ?

Comment (burgerdev) [reply packages/by-name/kata/kata-kernel-uvm/package.nix:11]: I think what I missed is that the base config files are different for the two kernels - sorry.

File: packages/by-name/kata/kata-kernel-uvm/package.nix (hunk 3 above)
Comment (msanft) [reply packages/by-name/kata/kata-kernel-uvm/package.nix:33]: These are from [upstream](https://github.com/kata-containers/kata-containers/blob/2c6126d3ab708e480b5aad1e7f7adbe22ffaa539/tools/packaging/kernel/configs/fragments/common/confidential_containers/cryptsetup.conf#L2). I'll add a comment indicating that.

**Review by burgerdev — APPROVED** [review]

**Pull request merged**